Using-the-Compiler-API.md

```diff
@@ -283,6 +283,224 @@ function reportWatchStatusChanged(diagnostic: ts.Diagnostic) {
 watchMain();
 ```
 
+## A minimal incremental compiler
+
+To compile incrementally, we need to create a builder program that reads the older program by reading tsbuildinfo file. This is as simple as calling `createIncrementalProgram`. `createIncrementalProgram` abstracts any interaction with the underlying system in the `IncrementalCompilerHost` interface. The `IncrementalCompilerHost` allows the compiler to read and write files, get the current directory, ensure that files and directories exist, and query some of the underlying system properties such as case sensitivity and new line characters. For convenience, we expose a function to create a default host using `createIncrementalCompilerHost`.
+
+```TypeScript
+import * as ts from "typescript";
+
+function incrementalCompile(): void {
+  const configPath = ts.findConfigFile(
+    /*searchPath*/ "./",
+    ts.sys.fileExists,
+    "tsconfig.json"
+  );
+  if (!configPath) {
+    throw new Error("Could not find a valid 'tsconfig.json'.");
+  }
+
+  const config = ts.getParsedCommandLineOfConfigFile(
+    configPath,
+    /*optionsToExtend*/ { incremental: true },
+    /*host*/ {
+      ...ts.sys,
+      onUnRecoverableConfigFileDiagnostic: d => console.error(ts.flattenDiagnosticMessageText(d, "\n"));
+    }
+  );
+  if (!config) {
+    throw new Error("Could not parse 'tsconfig.json'.");
+  }
+
+  const program = ts.createIncrementalProgram({
+    rootName: config.fileNames,
+    options: config.options,
+    configFileParsingDiagnostics: ts.getConfigFileParsingDiagnostics(config),
+    projectReferences: config.projectReferences
+    // createProgram can be passed in here to choose strategy for incremental compiler just like when creating incremental watcher program.
+    // Default is ts.createSemanticDiagnosticsBuilderProgram
+  });
+  const diagnostics = [
+    ...program.getConfigFileParsingDiagnostics(),
+    ...program.getSyntacticDiagnostics(),
+    ...program.getOptionsDiagnostics(),
+    ...program.getGlobalDiagnostics(),
+    ...program.getSemanticDiagnostics() // Get the diagnostics before emit to cache them in the buildInfo file.
+  ];
```

Comment:
We not `ts.getPreEmitDiagnostics()` instead of duplicating the `get___Diagnostics()` calls?

What about [`ts.sortAndDeduplicateDiagnostics(diagnostics)`](https://github.com/microsoft/TypeScript/blob/e934a0d1704e21dbecc2eddc2d232e15053107e6/src/compiler/program.ts#L335)?